src/Synthesizer.ml

```diff
@@ -233,16 +242,35 @@ let solve_impl (config : Config.t) (task : task) (stats : stats) =
     in f_divide (apply_cells [] arg_types) (List.length arg_types) op_level expr_level (cost - 1)
   in
   let expand_component op_level expr_level cost candidates component =
+
+    let unify_component (args : Expr.synthesized list) (comp : Expr.component) : Expr.component option =
+      let typs = List.map args ~f:(fun s -> Value.typeof s.outputs.(0)) in
+      let open TypeUnify in
+      let (cod, d) = match rename (collect_varnames typs) (comp.codomain :: comp.domain) with
+        | hd :: tl -> (hd, tl)
+        | _ -> raise (Internal_Exn "Impossible case!") in
+      let subst = unify typs d in
+      match subst with
+        | None -> None
+        | Some subst ->
+          let new_codomain = apply_subst subst cod in
+          Some {comp with codomain = new_codomain}
```

Comment:
I think modifying the codomain at this point might not be correct ... Basically, we have already picked the `candidates` array where we want to put the resulting expression (after applying the component). At this point, it's possible that we obtain a different codomain than the one that corresponds to the `candidates` array we had picked.

If I understand your changes correctly, you allow a codomain such as `TYPEDVAR(_)` to be mapped to _all_ candidate arrays. This makes sense since the type variable may bind to any concrete type after unification. However, at this point during the synthesis, we have already picked a candidate array to place the resulting expression into, so we should constraint the type variables in the domain and codomain accordingly and then applying to matching arguments.

Let me know what you think, and if I'm missing something here.

Reply:
Yes Saswat, I think you are correct in that this step is redundant. 
I realized that components with codomain being 'TYPEDVAR(_)' are tricky for us now as it requires a dynamically determined candidate array, but this is not intended to fix that -- to fix that we probably need to make structural changes, so my intention was to ask how you think first and leave that for the future. 

Similarly, for components with 'TYPEDVAR(_)' as an argument in its domain, the process of picking the 'typed_candidate' in 'apply_cells' would require concatenating all the other candidates at runtime --- I'm not sure if performing this concatenation again and again would be too costly, or should we cache it in some ways. Again, please let me know your thoughts on this.

Finally, to address why I was wrong, what I was thinking was this: to apply any component with type variables, we need an extra step on top of 'apply_cells' to validate if the component can be applied to the candidates we picked, via a unification step. I wrote this step in the function 'unify_component'. Now, to perform this unification, we need some way to extract type info from the candidates (of type 'synthesized'). There are two ways to do this: 1) take candidate.expr, and look at the component's codomain in FCall (i haven't thought about how to deal with Var/Const though, but there should be a similar way to look it up). If we use this approach, then we'd have to modify codomain to pass down the new type information after unification.

2) take candidate.outputs, and reconstruct the type from the actual values obtained after computation. Turned out this is what I used in 'unify_component', so this renders information from the candidate expr's FCall component's codomain useless. 

TL;DR: I was doing useless work for approach 1) while it turned out I implemented type unification with approach 2). So you're right, modifying codomain is not needed. I can change that first. Meanwhile, if you can let me know your thoughts on how best to deal with TYPEDVAR(_) in domain and codomain, I can make further changes to fully enable TYPEDVAR components.